File: content/snippets/python/interview_lcci_16_15_master_mind.py
from typing import List


def master_mind(solution: str, guess: str) -> List[int]:
    hits = 0
    cs = [0] * 26
    cg = [0] * 26
    for s, g in zip(solution, guess):
        if s == g:
            hits += 1
        else:
            cs[ord(s) - ord("A")] += 1
            cg[ord(g) - ord("A")] += 1
    pseudo = sum(min(cs[i], cg[i]) for i in range(26))
    return [hits, pseudo]


def main() -> None:
    print(master_mind("RGBY", "GGRR"))


if __name__ == "__main__":
    main()

File: content/snippets/cpp/interview_lcci_16_15_master_mind.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

vector<int> masterMind(const string& solution, const string& guess) {
    int hits = 0;
    vector<int> cs(26, 0), cg(26, 0);
    for (int i = 0; i < 4; ++i) {
        if (solution[i] == guess[i]) {
            ++hits;
        } else {
            ++cs[solution[i] - 'A'];
            ++cg[guess[i] - 'A'];
        }
    }
    int pseudo = 0;
    for (int i = 0; i < 26; ++i) {
        pseudo += min(cs[i], cg[i]);
    }
    return {hits, pseudo};
}

int main() {
    vector<int> ans = masterMind("RGBY", "GGRR");
    cout << "[" << ans[0] << ", " << ans[1] << "]" << endl;
    return 0;
}
